File: category-backend/src/main/java/com/namit/categorybackend/category/service/CategoryService.java
package com.namit.categorybackend.category.service;

import com.namit.categorybackend.category.dto.CategoryRequest;
import com.namit.categorybackend.category.dto.CategoryResponse;
import com.namit.categorybackend.common.response.PagedResponse;
import org.springframework.data.domain.Page;

public interface CategoryService {

    // Creates a new category after validating uniqueness
    CategoryResponse createCategory(CategoryRequest request);

    // Gets all categories
    PagedResponse<CategoryResponse> getAllCategories(int page, int size, String status);

    // Retrieves an active category by its ID.
    CategoryResponse getCategoryById(Long id);

    // Updates the category
    CategoryResponse updateCategory(Long id, CategoryRequest request);

    // Performs soft delete by marking category as inactive, reassigning products
    void deleteCategory(Long id, Long reassignCategoryId);

    // Toggles the category status (Active <-> Inactive), reassigning products on
    // deactivation
    CategoryResponse toggleCategoryStatus(Long id, Long reassignCategoryId);

    // Returns count of products in a category
    long getProductCount(Long categoryId);
}

File: category-frontend/src/services/categoryService.js
import api from "../api/axios";

export const getCategories = (page = 0, size = 10, status = "active") =>
    api.get(`/categories?page=${page}&size=${size}&status=${status}`);

export const getCategoryById = (id) => api.get(`/categories/${id}`);

export const createCategory = (data) => api.post("/categories", data);

export const updateCategory = (id, data) => api.put(`/categories/${id}`, data);

export const deleteCategory = (id, reassignCategoryId = null) =>
    api.delete(`/categories/${id}`, {
        data: { reassignCategoryId },
    });

export const toggleCategoryStatus = (id, reassignCategoryId = null) =>
    api.patch(`/categories/${id}/toggle`, { reassignCategoryId });

export const getCategoryProductCount = (id) =>
    api.get(`/categories/${id}/product-count`);
